src/ORM/Query.php

```diff
@@ -370,30 +370,167 @@ public function eagerLoader(EagerLoader $instance = null)
+     *
+     * // For an author query, load his region, state and country
+     * $query->setContain('Regions.States.Countries');
+     * ```
+     *
+     * It is possible to control the conditions and fields selected for each of the
+     * contained associations:
+     *
+     * ### Example:
+     *
+     * ```
+     * $query->setContain(['Tags' => function ($q) {
+     *     return $q->where(['Tags.is_popular' => true]);
+     * }]);
+     *
+     * $query->setContain(['Products.Manufactures' => function ($q) {
+     *     return $q->select(['name'])->where(['Manufactures.active' => true]);
+     * }]);
+     * ```
+     *
+     * Each association might define special options when eager loaded, the allowed
+     * options that can be set per association are:
+     *
+     * - `foreignKey`: Used to set a different field to match both tables, if set to false
+     *   no join conditions will be generated automatically. `false` can only be used on
+     *   joinable associations and cannot be used with hasMany or belongsToMany associations.
+     * - `fields`: An array with the fields that should be fetched from the association.
+     * - `finder`: The finder to use when loading associated records. Either the name of the
+     *   finder as a string, or an array to define options to pass to the finder.
+     * - `queryBuilder`: Equivalent to passing a callable instead of an options array.
+     *
+     * ### Example:
+     *
+     * ```
+     * // Set options for the hasMany articles that will be eagerly loaded for an author
+     * $query->setContain([
+     *     'Articles' => [
+     *         'fields' => ['title', 'author_id']
+     *     ]
+     * ]);
+     * ```
+     *
+     * Finders can be configured to use options.
+     *
+     * ```
+     * // Retrieve translations for the articles, but only those for the `en` and `es` locales
+     * $query->setContain([
+     *     'Articles' => [
+     *         'finder' => [
+     *             'translations' => [
+     *                 'locales' => ['en', 'es']
+     *             ]
+     *         ]
+     *     ]
+     * ]);
+     * ```
+     *
+     * When containing associations, it is important to include foreign key columns.
+     * Failing to do so will trigger exceptions.
+     *
+     * ```
+     * // Use special join conditions for getting an Articles's belongsTo 'authors'
+     * $query->setContain([
+     *     'Authors' => [
+     *         'foreignKey' => false,
+     *         'queryBuilder' => function ($q) {
+     *             return $q->where(...); // Add full filtering conditions
+     *         }
+     *     ]
+     * ]);
+     * ```
+     *
+     * @param array|string $associations List of table aliases to be queried.
+     * @param bool $override Whether override previous list with the one passed
+     * defaults to merging previous list with the new one.
+     * @return $this
+     */
+    public function setContain($associations, $override = false)
```

Comment:
The query builder is one of the few places where set prefixes become a bit clunky. If all the methods like where(), order(), and group() all had set  prefixes I think the interface would be less pleasant to use. However, I think adding get/clear methods makes sense. Is there a way we could deprecate the other modes from `contain()` and add the get/clear methods?